File: NeetCode150/Stack/ValidParentheses/ValidParentheses.py
"""
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
An input string is valid if:
Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.

Input: s = "()"
Output: true
"""

#Simple solution using stack (by xiaoying10101)

class Solution(object):
    def isValid(self, s):
        stack = []
        dict = {"]":"[", "}":"{", ")":"("}
        for char in s:
            if char in dict.values():
                stack.append(char)
            elif char in dict.keys():
                if stack == [] or dict[char] != stack.pop():
                    return False
            else:
                return False
        return stack == []
      
"""
This solution can be solved using stack to track every open parentheses and store them in an array. If we meet a close parentheses, check if the last open parentheses
is match or if stack is not empty. If false then the string is invalid because it might has single open parentheses in between different parentheses or string contain
only closing parentheses/open parentheses.
"""
       

File: NeetCode150/Stack/ValidParentheses/ValidParentheses.java
package NeetCode150.Stack.ValidParentheses;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<Character>();

        Map<Character, Character> parentheses = Map.of(
            '(', ')',
            '{', '}',
            '[', ']'
        );

        for (char letter : s.toCharArray()) {
            if (parentheses.containsKey(letter)) {
                stack.push(letter);
                continue;
            }

            if (stack.isEmpty() || !parentheses.get(stack.getFirst()).equals(letter)) {
                return false;
            }

            stack.pop();
        }

        return stack.isEmpty();
    }
}

// In this problem we can achieve optimal solution by using stack data structure using Deque Array.
// Basically, we can iterate throught the string and add the char to the stack if its an open parentheseses.
// If its a close parentheses, you can check with first element of the stack (lattest added). If its a matching type, remove it.
// The closing parentheses ALWAYS has to match with the latest added open parentheses. In the end, the stack will be empty which mean its a valid parentheses.
// This solution has O(n) time complexity (all the ltter in string) and O(n) space complexity (for storing open parentheses)

